### DISCUSSION TITLE
AIOBLE :strange beavior if used with Pi pico w or ESP32 c3

### DISCUSSION THREAD

**Discussion by BOB63** [micropython/micropython#15857] [ESP32] [answered]
I found this sketch that permit to connect two Pi Pivo W via BLE and exchange messages between a Ble Server and a Ble Client including also an  acknoledge msg from the Client to the Server: (https://github.com/kevinmcaleer/pi_to_pico_bluetooth/blob/main/lesson_01/pico_a.py).
Used with the Pi Pico w it works perfectly but when I try to use with ESP32 C3 as Peripheral I got a problem .

The pico_a.py can be customised to works as client or serve  changing only the variable at line 12  " IAM = 'Peripheral' or IAM = 'Client' "
When I use the sketch with an ESP32c3 as Client there are no problems  but when I use configured as Peripheral I got this error 

```
MPY: soft reboot
Peripheral starting to advertise
Traceback (most recent call last):
  File "<stdin>", line 198, in <module>
  File "asyncio/core.py", line 1, in run
  File "asyncio/core.py", line 1, in run_until_complete
  File "asyncio/core.py", line 1, in run_until_complete
  File "<stdin>", line 196, in main
  File "asyncio/funcs.py", line 1, in gather
  File "asyncio/core.py", line 1, in run_until_complete
  **File "<stdin>", line 118, in run_peripheral_mode**
  **File "/lib/aioble/peripheral.py", line 158, in advertise**
OSError: -530
```

The same sketch loaded on the Pi Pico w works fine.
About the " OSError -530 " I'mk not able to find any explanation in google.
The code line 118 of pico_a.py seems nothing special ( services=[BLE_SVC_UUID], )  
and common to other sketch that I've found for ESP32.
```
async def run_peripheral_mode():
    """ Run the peripheral mode """

    # Set up the Bluetooth service and characteristic
    ble_service = aioble.Service(BLE_SVC_UUID)
    characteristic = aioble.Characteristic(
        ble_service,
        BLE_CHARACTERISTIC_UUID,
        read=True,
        notify=True,
        write=True,
        capture=True,
    )
    aioble.register_services(ble_service)

    print(f"{BLE_NAME} starting to advertise")

    while True:
        async with await aioble.advertise(
            BLE_ADVERTISING_INTERVAL,
            name="BLE_NAME",
            services=[BLE_SVC_UUID],
            appearance=BLE_APPEARANCE) as connection:
            print(f"{BLE_NAME} connected to another device: {connection.device}")

            tasks = [
                asyncio.create_task(send_data_task(connection, characteristic)),
            ]
            await asyncio.gather(*tasks)
            print(f"{IAM} disconnected")
            break
```

At this point I don't know if the issue could be inside the aioble module **"/lib/aioble/peripheral.py", line 158** 
and if aioble modules change in such way if loaded on ESP32 or on Pi Pico.

Any suggestion is wellcome.

**Comment by bidrohini** [comment]
The ESP32-C3 has a slightly different BLE stack than the Pi Pico W. It’s possible that the aioble module (which is often based on the ubluetooth or other BLE libraries) handles advertising differently on these platforms.

**Comment by andrewleech** [comment]
I just tried running the same code on a C3 and ran into the same issue, there's certainly something wrong there.
the esp chips all use a nimble based stack, where pico uses btstack - but aioble is intended to work the same on both. The pyboard-d also uses nimble, so most of the official testing during development would have been done with that iirc.

As a point of comparison I just ran the aioble test sensor example on the C3 and it _did_ run correctly - I think from a quick look it's quite similar to `pico_a.py` in terms of how the peripheral use is configured at least, maybe you could try copying bits back and forth between both examples to figure out which bit of difference it causing the failure: https://github.com/micropython/micropython-lib/blob/master/micropython/bluetooth/aioble/examples/temp_sensor.py

**Comment by BOB63** [comment, marked as answer]
@andrewleech ,
I've found the problem comparing how are set the advertising interval in pico_a.py vs temp_sensor_.py , because the syntax of advertising code is almost the same in both programs  .

I've verified that with the ESP32_C3 setting the advertising interval in pico_a.py above 600 fails with the original error while a value below works fine .

Unfortunately for some weeks I will not have no possibility to verify if this advertise interval setting works fine also on pi picow.

What I've not clear is the format used in temp_sernsor.py to set advertise interval .
How is interpreted _ADV_INTERVAL_MS = 250_000  ?
250 ms ?
Is the first time that I see these format . 

tks

**Reply by andrewleech** [reply]
Good finding @BOB63 I wonder why the longer intervals don't work there, might need some deeper investigation. 

The number format is a python thing, you can add `_` into a number to make them more readable, like a comma is often used to show thousands. eg. 10,000 

When the number is read the underscore is ignored.

```
>>> 250_000
250000
>>>
```

In the specific case, `_ADV_INTERVAL_MS = 250_000` will mean 250 seconds.

**Comment by BOB63** [comment]
@andrewleech 
This open another strange beavior because during the test I've set the adv interval at 250_000 and was working !
Now I've teste other value >600 and fails until iI set values above 20000 and then i can increase above  300000 without issues.
It sound really strange why the vale between 600 and 19999 continue to give the same error .
There isn't logic at all ...
